kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java

```diff
@@ -0,0 +1,228 @@
+                    // this should be impossible, because if any of the futures completed exceptionally then the thenApply()
+                    // won't be called
+                    throw new IllegalStateException(e);
+                }
+
+            }).toList();
+        });
+    }
+
+    private CompletableFuture<Map<String, DekContext<K>>> keyContexts(Map<String, K> topicToKekId) {
+        Map<K, List<String>> inverted = new HashMap<>();
+        Set<K> kekIds = new HashSet<>(topicToKekId.size());
+        topicToKekId.forEach((topicName, kekId) -> {
+            kekIds.add(kekId);
+            inverted.computeIfAbsent(kekId, k -> new ArrayList<>()).add(topicName);
+        });
+        var futures = kekIds.stream().map(kekId -> dekCache.forKekId(kekId).toCompletableFuture()).toList();
+
+        return join(futures).thenApply(list -> {
+            Map<String, DekContext<K>> result = new HashMap<>(topicToKekId.size());
+            list.forEach(dekContext -> inverted.get(dekContext.kekId()).forEach(topicName -> result.put(topicName, dekContext)));
+            return result;
+        });
+    }
+
+    @Override
+    public CompletionStage<RequestFilterResult> onProduceRequest(short apiVersion, RequestHeaderData header, ProduceRequestData request, FilterContext context) {
+        var topicNames = request.topicData().stream().map(TopicProduceData::name).collect(Collectors.toSet());
+        return kekSelector.selectKek(topicNames)
+                .thenCompose(kekMap -> keyContexts(kekMap).thenApply(topicNameToKeyContext -> {
+                    request.topicData().forEach(td -> {
+                        var keyContext = topicNameToKeyContext.get(td.name());
+                        if (keyContext != null) {
+                            td.setPartitionData(encryptPartition(keyContext, td.partitionData()));
+                        }
+                    });
+                    return request;
+                }).thenCompose(yy -> context.forwardRequest(header, request)));
+    }
+
+    private List<PartitionProduceData> encryptPartition(DekContext<K> dekContext,
+                                                        List<PartitionProduceData> oldPartitionData) {
+        List<PartitionProduceData> newPartitionData = new ArrayList<>(oldPartitionData.size());
+        for (var pd : oldPartitionData) {
+            int partitionId = pd.index();
+            ByteBuffer buffer = null; /// TODO determine how we should allocate this (is it a configuration parameter??)
+            MemoryRecords records = (MemoryRecords) pd.records();
+            MemoryRecordsBuilder builder = recordsBuilder(buffer, records);
+            for (var kafkaRecord : records.records()) {
+                ByteBuffer output = null;
+                try {
+                    output = bufferPool.acquire(dekContext.encodedSize(kafkaRecord.valueSize()));
+                    dekContext.encode(kafkaRecord.value(), output);
```

Comment:
should we have the option to compress the encrypted bytes?  probably something we can defer.